### PR TITLE
Add post type- and taxonomy-specific registration hooks

### PR DESCRIPTION
**Pull request by dlh01**
Trac ticket: https://core.trac.wordpress.org/ticket/53212

### CODE REVIEW

File: src/wp-includes/class-wp-post-type.php
Comment (pbiron) [inline src/wp-includes/class-wp-post-type.php]: I think it would better if the `post_type`-specific filter were applied **before** `register_post_type_args` .

The same goes for the `taxonomy`-specific filter...and both new `post_type` and `taxonomy`-specific actions.

Comment (pbiron) [reply src/wp-includes/class-wp-post-type.php]: I have a slight preference for the name of the new filters to be `register_post_type_XXXX_args`, but won't object if other folk like them this way

Comment (pbiron) [reply src/wp-includes/class-wp-post-type.php]: I do feel more strongly, tho, that where ever the post type name appears in the name of the filter, it should be done as:

```
$post_type = $this->name;
apply_filters( "register_{$post_type}_post_type_args", ... );
// or apply_filters( "register_post_type_{$post_type}_args", ... );
```

Doing that makes the Code Reference cleaner (there has been a lot of that kind of cleanup for similar filters/actions recently).

**Review by pbiron — CHANGES_REQUESTED** [review]
Other than the comments I've made about the naming/position of the new filters/actions, this looks good to me.

In particular, I think the unit tests look find.

**Comment by dlh01** [comment]
Thanks for the review, @pbiron!

> I think it would better if the post_type-specific filter were applied before register_post_type_args .

I searched for a pattern to follow in existing hooks but couldn't find much consistency. `save_post_{$post->post_type}` comes before `save_post`, but `edited_term` comes before `edited_{$taxonomy}`...etc.

My thinking was that adding the new hooks after the existing hooks would preserve the behavior of those existing hooks, framing the new hooks as customization points that "add-on" to the current logic. But I suppose I don't have much of a preference here. Happy to defer to the judgment of the component maintainers and committers.

> I have a slight preference for the name of the new filters to be register_post_type_XXXX_args, but won't object if other folk like them this way

To me, `register_post_type_XXXX_args` sounds like it could also be a filter for a specific registration argument, like `register_post_type_rewrite_args` or something, whereas `register_XXXX_post_type_args` puts the focus on "which post type?" So, I have a slight preference for the current approach. But maybe there's also an alternate reading of the current approach in the PR, too, that I'm not hearing 😄 

> I do feel more strongly, tho, that where ever the post type name appears in the name of the filter, it should be done as:
> 
> ```
> $post_type = $this->name;
> apply_filters( "register_{$post_type}_post_type_args", ... );
> // or apply_filters( "register_post_type_{$post_type}_args", ... );
> ```
> 
> Doing that makes the Code Reference cleaner (there has been a lot of that kind of cleanup for similar filters/actions recently).

Good to know! Updated in the latest commit.

**Review by peterwilsoncc — COMMENTED** [review]
Wow! WordPress is quite inconsistent with `do_{$post_type}_thing` vs `do_thing_{$post_type}`.

It seems the former is in the slight majority so the names on this PR seem fine. 

Wow! WordPress is quite inconsistent as to whether `do_{$post_type}_thing` is placed before or after `do_thing`.

After seems more consistent so I'm happy with that too. 

Thanks for including unit tests.

**Comment by peterwilsoncc** [comment]
Merged in https://core.trac.wordpress.org/changeset/53126, thanks.

File: src/wp-includes/class-wp-post-type.php
Comment (pbiron) [inline src/wp-includes/class-wp-post-type.php:448]: Another thing I just noticed: this description in the DocBlock should read:

> The dynamic portion of the filter name, `$post_type`, refers to the post type key.

Same for the one on line 323 in `class-wp-taxonomy.php` (where `$post_type` is, of course, replaced by `$taxonomy`).  Sorry for not catching those in my earlier review.

Comment (dlh01) [reply src/wp-includes/class-wp-post-type.php:448]: My bad for missing those when I swapped in those variables. I'll create a follow-up PR later today.

Comment (pbiron) [reply src/wp-includes/class-wp-post-type.php:448]: thanx.  It's also recently become standard to add something like the following to the DocBlock (after the `The dynamic portion of the filter name...`):

>  Possible hook names include:
>    - `register_post_post_type_args`
>   - `register_page_post_type_args`

i.e., listing the "expanded" filter name with built-in post types (or taxonomies).  For an example, see the Code Reference entry for [manage_taxonomies_for_post_type_columns](https://developer.wordpress.org/reference/hooks/manage_taxonomies_for_post_type_columns/)

**Pull request closed without merging**
